frontend/src/components/Table.tsx

```diff
@@ -0,0 +1,115 @@
+import { Pagination, Toolbar, ToolbarContent, ToolbarItem } from '@patternfly/react-core';
+import {
+  TableComposable,
+  Thead,
+  Tr,
+  Th,
+  Tbody,
+  TableComposableProps,
+  Caption,
+  Td,
+} from '@patternfly/react-table';
+import React, { useEffect } from 'react';
+import useTableColumnSort, { SortableData } from '~/utilities/useTableColumnSort';
+
+type TableProps<DataType> = {
+  data: DataType[];
+  columns: SortableData<DataType>[];
+  rowRenderer: (data: DataType, rowIndex: number) => React.ReactNode;
+  enablePagination?: boolean;
+  minPageSize?: number;
+  toolbarContent?: React.ReactElement<typeof ToolbarItem>;
+  emptyTableView?: React.ReactElement<typeof Tr>;
+  caption?: string;
+} & Omit<TableComposableProps, 'ref' | 'data'>;
+
+export const Table = <T,>({
+  data,
+  columns,
+  rowRenderer,
+  enablePagination,
+  minPageSize = 10,
+  toolbarContent,
+  emptyTableView,
+  caption,
+  ...props
+}: TableProps<T>): React.ReactElement => {
+  const [page, setPage] = React.useState(1);
+  const [pageSize, setPageSize] = React.useState(minPageSize);
+
+  // update page to 1 if data changes (common when filter is applied)
+  useEffect(() => setPage(1), [data]);
+
+  const sort = useTableColumnSort<T>(columns, 0);
+
+  const showPagination = enablePagination && data.length > minPageSize;
+  const pagination = (pageDirection: 'up' | 'down') =>
+    showPagination && (
+      <Pagination
+        dropDirection={pageDirection}
+        perPageComponent="button"
+        itemCount={data.length}
+        perPage={pageSize}
+        page={page}
+        onSetPage={(e, newPage) => setPage(newPage)}
+        onPerPageSelect={(e, newSize, newPage) => {
+          setPageSize(newSize);
+          setPage(newPage);
+        }}
+        widgetId="table-pagination"
+      />
+    );
+
+  return (
+    <>
+      <Toolbar>
+        <ToolbarContent>
+          {toolbarContent}
+          <ToolbarItem variant="pagination" alignment={{ default: 'alignRight' }}>
+            {pagination('down')}
+          </ToolbarItem>
+        </ToolbarContent>
+      </Toolbar>
+      <TableComposable {...props}>
+        {caption && <Caption>{caption}</Caption>}
+        <Thead>
+          <Tr>
+            {columns.map((col, i) => (
+              <Th
+                key={col.field + i}
```

Comment:
Some tables used the index

Reply:
Ah, okay -- don't think it's needed here. Use unique headers. This is likely a problem with your empty `field` states you have in all the various columns. `field: ''` is probably invalid... but TS doesn't have length tests for us to use.

Reply:
Its all the BYON notebook tables

Reply:
Maybe you revert this effort then -- if you can't really properly support that table, we can do it later. Let us not compromise the base table implementation because of old code.

Reply:
We should swing back in here and remove the index when we can make sure the field is not an empty string -- might go into the `useTableSortColumn` function to throw an error.